spec/controllers/admin/products_controller_spec.rb

```diff
@@ -0,0 +1,119 @@
+      get :new, {}, valid_session
+      expect(assigns(:product)).to be_a_new(Product)
+    end
+  end
+
+  describe "GET #edit" do
+    it "assigns the requested product as @product" do
+      product = Product.create! valid_attributes
+      get :edit, {:id => product.to_param}, valid_session
+      expect(assigns(:product)).to eq(product)
+    end
+  end
+
+  describe "POST #create" do
+    context "with valid params" do
+      it "creates a new Product" do
+        expect {
+          post :create, {:product => valid_attributes}, valid_session
+        }.to change(Product, :count).by(1)
+      end
+
+      it "assigns a newly created product as @product" do
+        post :create, {:product => valid_attributes}, valid_session
+        expect(assigns(:product)).to be_a(Product)
+        expect(assigns(:product)).to be_persisted
+      end
+
+      it "redirects to the product list" do
+        post :create, {:product => valid_attributes}, valid_session
+        expect(response).to redirect_to(admin_products_path)
+      end
+    end
+
+    context "with invalid params" do
+      it "assigns a newly created but unsaved product as @product" do
+        post :create, {:product => invalid_attributes}, valid_session
+        expect(assigns(:product)).to be_a_new(Product)
+      end
+
+      it "re-renders the 'new' template" do
+        post :create, {:product => invalid_attributes}, valid_session
+        expect(response).to render_template("new")
+      end
+    end
+  end
+
+  describe "PUT #update" do
+    context "with valid params" do
+      let(:new_attributes) {
+        {name: "Bar", price: 5.50 }
+      }
+
+      it "updates the requested product" do
+        product = Product.create! valid_attributes
+        put :update, {:id => product.to_param, :product => new_attributes}, valid_session
+        product.reload
+        expect(product.name).to eq("Bar")
+        expect(product.price).to eq(5.50)
+      end
+
+      it "assigns the requested product as @product" do
+        product = Product.create! valid_attributes
+        put :update, {:id => product.to_param, :product => valid_attributes}, valid_session
+        expect(assigns(:product)).to eq(product)
+      end
+
+      it "redirects to the product" do
+        product = Product.create! valid_attributes
+        put :update, {:id => product.to_param, :product => valid_attributes}, valid_session
+        expect(response).to redirect_to(admin_products_path)
+      end
+    end
+
+    context "with invalid params" do
+      it "assigns the product as @product" do
+        product = Product.create! valid_attributes
+        put :update, {:id => product.to_param, :product => invalid_attributes}, valid_session
+        expect(assigns(:product)).to eq(product)
+      end
+
+      it "re-renders the 'edit' template" do
+        product = Product.create! valid_attributes
+        put :update, {:id => product.to_param, :product => invalid_attributes}, valid_session
```

Comment:
Line is too long. [93/80]<br>Use the new Ruby 1.9 hash syntax.<br>Space inside { missing.<br>Space inside } missing.